hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/OrphanKeyDetectionTask.java

```diff
@@ -0,0 +1,154 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.ozone.recon.tasks;
+
+import com.google.inject.Inject;
+import org.apache.commons.lang3.tuple.ImmutablePair;
+import org.apache.commons.lang3.tuple.Pair;
+import org.apache.hadoop.hdds.utils.db.DBStore;
+import org.apache.hadoop.hdds.utils.db.Table;
+import org.apache.hadoop.ozone.om.OMMetadataManager;
+import org.apache.hadoop.ozone.om.helpers.OmKeyInfo;
+import org.apache.hadoop.ozone.om.helpers.WithParentObjectId;
+import org.apache.hadoop.ozone.recon.api.types.OrphanKeyMetaData;
+import org.apache.hadoop.ozone.recon.spi.ReconNamespaceSummaryManager;
+import org.apache.hadoop.ozone.recon.spi.impl.ReconDBProvider;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+
+import static org.apache.hadoop.ozone.om.OmMetadataManagerImpl.DELETED_DIR_TABLE;
+import static org.apache.hadoop.ozone.recon.spi.impl.ReconDBDefinition.ORPHAN_KEYS_METADATA;
+
+/**
+ * Task class to iterate over the OM DB and detect the orphan keys metadata.
+ */
+public class OrphanKeyDetectionTask implements ReconOmTask {
+
+  private static final Logger LOG =
+      LoggerFactory.getLogger(OrphanKeyDetectionTask.class);
+  private DBStore reconDbStore;
+  private ReconNamespaceSummaryManager reconNamespaceSummaryManager;
+  private final Table<Long, OrphanKeyMetaData> orphanKeysMetaDataTable;
+
+  @Inject
+  public OrphanKeyDetectionTask(
+      ReconDBProvider reconDBProvider,
+      ReconNamespaceSummaryManager reconNamespaceSummaryManager)
+      throws IOException {
+    this.reconDbStore = reconDBProvider.getDbStore();
+    this.reconNamespaceSummaryManager = reconNamespaceSummaryManager;
+    this.orphanKeysMetaDataTable =
+        ORPHAN_KEYS_METADATA.getTable(reconDbStore);
+  }
+
+  @Override
+  public String getTaskName() {
+    return "OrphanKeyDetectionTask";
+  }
+
+  public Collection<String> getTaskTables() {
+    List<String> taskTables = new ArrayList<>();
+    taskTables.add(DELETED_DIR_TABLE);
```

Comment:
This is used for cleanup orphan parent if already exist in deletedDirTable for FSO. So based on logic, we need cleanup with below task,
1. bucket delete event
Also can help in handling nssummary memory leak where objectId is never cleaned for DELETE event of directory and bucket
- We can rename this class as used for nssummary or can be merged with existing NSSummaryTaskDbEvents.

Reply:
Done.